libs/server-sdk/src/evaluation/rules.cpp

```diff
@@ -0,0 +1,215 @@
+            continue;
+        }
+
+        auto maybe_contains =
+            Contains(*segment_ptr->item, context, store, stack);
+
+        if (!maybe_contains) {
+            return tl::make_unexpected(maybe_contains.error());
+        }
+
+        if (maybe_contains.value()) {
+            return MaybeNegate(clause, true);
+        }
+    }
+
+    return MaybeNegate(clause, false);
+}
+
+tl::expected<bool, Error> MatchNonSegment(
+    Clause const& clause,
+    launchdarkly::Context const& context) {
+    if (!clause.attribute.Valid()) {
+        return tl::make_unexpected(
+            Error::InvalidAttributeReference(clause.attribute.RedactionName()));
+    }
+
+    if (clause.attribute.IsKind()) {
+        for (auto const& clause_value : clause.values) {
+            for (auto const& kind : context.Kinds()) {
+                if (operators::Match(clause.op, kind, clause_value)) {
+                    return MaybeNegate(clause, true);
+                }
+            }
+        }
+        return MaybeNegate(clause, false);
+    }
+
+    Value const& attribute = context.Get(clause.contextKind, clause.attribute);
+    if (attribute.IsNull()) {
+        return false;
+    }
+
+    if (attribute.IsArray()) {
+        for (Value const& clause_value : clause.values) {
+            for (Value const& context_value : attribute.AsArray()) {
+                if (operators::Match(clause.op, context_value, clause_value)) {
+                    return MaybeNegate(clause, true);
+                }
+            }
+        }
+        return MaybeNegate(clause, false);
+    }
+
+    if (std::any_of(clause.values.begin(), clause.values.end(),
+                    [&](Value const& clause_value) {
+                        return operators::Match(clause.op, attribute,
+                                                clause_value);
+                    })) {
+        return MaybeNegate(clause, true);
+    }
+
+    return MaybeNegate(clause, false);
+}
+
+tl::expected<bool, Error> Contains(Segment const& segment,
+                                   Context const& context,
+                                   data_store::IDataStore const& store,
+                                   detail::EvaluationStack& stack) {
+    auto guard = stack.NoticeSegment(segment.key);
+    if (!guard) {
+        return tl::make_unexpected(Error::CyclicSegmentReference(segment.key));
+    }
+
+    if (segment.unbounded) {
+        // TODO(sc209881): set big segment status to NOT_CONFIGURED.
+        return false;
+    }
+
+    if (IsTargeted(context, segment.included, segment.includedContexts)) {
+        return true;
+    }
+
+    if (IsTargeted(context, segment.excluded, segment.excludedContexts)) {
+        return false;
+    }
+
+    for (auto const& rule : segment.rules) {
+        if (!segment.salt) {
+            return tl::make_unexpected(Error::MissingSalt(segment.key));
+        }
+        if (Match(rule, context, store, stack, segment.key, *segment.salt)) {
```

Comment:
Is this correct? Reflexively it feels incorrect, but maybe I am not following the logic correctly (I would expect the value to be used, versus presence of value?).

Reply:
Yes, this is wrong. I missed that this was returning a `tl::expected`. Updated and added a unit test to make sure this piece is getting exercised.